### PR TITLE
apply fix suggested by lcnr

### PR DESCRIPTION
**Pull request by folkertdev**

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/collect/type_of.rs [hunk 1]
```diff
@@ -46,13 +60,57 @@ fn anon_const_type_of<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> Ty<'tcx> {
...
-                | hir::InlineAsmOperand::SymFn { anon_const } => anon_const.hir_id == hir_id,
-                _ => false,
-            }) =>
+            if let Some((anon_const, op_sp)) = asm.operands.iter().find_map(find_sym_fn) =>
         {
-            tcx.typeck(def_id).node_type(hir_id)
+            let ty = tcx.typeck(def_id).node_type(hir_id);
+
+            match ty.kind() {
+                ty::Never | ty::Error(_) => ty,
```
Comment (Amanieu) [inline compiler/rustc_hir_analysis/src/collect/type_of.rs]: It occurs to me that we probably don't want to allow `ty::Never` for `sym` operands, that may cause issues later.

File: compiler/rustc_hir_analysis/src/collect/type_of.rs [hunk 2]
```diff
@@ -46,13 +60,57 @@ fn anon_const_type_of<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> Ty<'tcx> {
...
+                            tcx.def_span(anon_const.def_id),
+                            format!("is {} `{}`", ty.kind().article(), ty),
+                        )
+                        .with_help("`sym` operands must refer to either a function or a static")
+                        .emit();
+
+                    Ty::new_error_with_message(
+                        tcx,
+                        span,
+                        format!("invalid type for `const` operand"),
```
Comment (Amanieu) [inline compiler/rustc_hir_analysis/src/collect/type_of.rs]:
```suggestion
                        format!("invalid type for `sym` operand"),
```

File: compiler/rustc_hir_analysis/src/collect/type_of.rs (hunk 1 above)
Comment (folkertdev) [inline compiler/rustc_hir_analysis/src/collect/type_of.rs]:
```suggestion
                ty::Error(_) => ty,
```

File: compiler/rustc_hir_analysis/src/check/intrinsicck.rs [hunk 3]
```diff
@@ -455,48 +455,22 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
...
-                                )
-                                .with_help(
-                                    "`sym` operands must refer to either a function or a static",
-                                )
-                                .emit();
-                        }
-                    };
+                    debug_assert!(matches!(
+                        self.tcx.type_of(anon_const.def_id).instantiate_identity().kind(),
+                        ty::Error(_) | ty::Never | ty::FnDef(..)
```
Comment (folkertdev) [inline compiler/rustc_hir_analysis/src/check/intrinsicck.rs]:
```suggestion
                        ty::Error(_) | ty::FnDef(..)
```

**Comment by folkertdev** [comment]
all fixed, you can squash the changes as you merge this branch into yours

**Pull request merged**